Preserve column order after unnesting

aaronwolen opened the pull request:
I run into this issue frequently enough that I thought I'd take a crack at it. The proposed change is pretty straightforward, so much so that I wonder if I'm overlooking something—especially in light of #145's conclusion. But I haven't found a case where it breaks down and all tests pass after adjusting a few that assumed the columns would be reordered.

hadley commented on tests/testthat/test-unnest.R:
```diff
@@ -162,3 +162,8 @@ test_that("grouping is preserved", {
+
+test_that("column order is preserved", {
+  df <- tibble(x = "a", y = list("b", "c"), z = "d")
+  expect_equal(df %>% unnest() %>% names(), c("x", "y", "z"))
```
You can use `expect_named()` here

hadley commented on tests/testthat/test-unnest.R:
```diff
@@ -162,3 +162,8 @@ test_that("grouping is preserved", {
   expect_equal(dplyr::groups(df), dplyr::groups(rs))
 })
+
+test_that("column order is preserved", {
```
Could you add another test for the case where you are unnesting a list of data frames?

aaronwolen commented:
Cool, I just implemented your suggestions and updated the NEWS.

aaronwolen commented:
Also updated the existing tests to use `expect_named()` where appropriate and fixed a typo. I thought a single PR would be easier for you but I'm happy to split these up if that's preferable.

krlmlr commented on tests/testthat/test-unnest.R:
```diff
@@ -167,3 +167,11 @@ test_that("grouping is preserved", {
+  df <- tibble(x = "a", y = list("b", "c"), z = "d")
+  expect_named(unnest(df), c("x", "y", "z"))
+
+  df$y <- list(tibble(y = "b"), tibble(y = "c"))
```
The test would be clearer if the internal tibbles had two columns with different names than `y` .

aaronwolen replied:
I see what you mean. What do you think about:

```r
df$y <- list(tibble(y = "b", y2 = "e"), tibble(y = "c", y2 = "f"))
```

This verifies the order of the original columns (`x`, `y`, and `z`) is preserved and new columns from the internal tibbles (`y2`) are appended.

```r
unnest(df)
## # A tibble: 2 x 4
##   x     y     z     y2   
##   <chr> <chr> <chr> <chr>
## 1 a     b     d     e    
## 2 a     c     d     f 
```

krlmlr replied:
I'd have expected the new column order to be x, y, y2, z, same if y isn't part of the nested tibble. What's the rationale behind leaving only y in the original place?

hadley commented:
@aaronwolen would you be interested in finishing this off for the next version of tidyr? If I don't hear back from you in a week, I'll assume that you're busy with other stuff, and I'll take it over the finish line.

aaronwolen commented:
Absolutely. I'll incorporate @krlmlr's suggestion re column ordering and push an update this week.

aaronwolen commented:
Accounting for the order of variables within nested data frames made this fix _slightly_ more complicated 😬

hadley commented on tests/testthat/test-unnest.R:
```diff
@@ -154,21 +154,21 @@ test_that("unnest drops list cols if expanding", {
   out <- df %>% unnest(y)
 
-  expect_equal(names(out), c("x", "z", "y"))
+  expect_equal(class(out$z), "list")
```
I think you should preserve the original test here.

hadley commented on tests/testthat/test-unnest.R:
```diff
@@ -154,21 +154,21 @@ test_that("unnest drops list cols if expanding", {
 
-  expect_equal(df %>% unnest(y, .drop = TRUE) %>% names(), c("x", "y"))
-  expect_equal(df %>% unnest(z, .drop = FALSE) %>% names(), c("x", "y", "z"))
+  expect_named(df %>% unnest(y, .drop = TRUE), c("x", "y"))
```
Since there's now only a single `%>%`, I think you should just get rid of it too.

aaronwolen commented:
Thanks! I just included your suggestions and updated the news. 

(Loving the new pivot fxns, btw!)

hadley commented on R/unnest.R:
```diff
@@ -157,7 +157,20 @@ unnest.data.frame <- function(data, ..., .drop = NA, .id = NULL,
   out <- dplyr::bind_cols(rest, unnested_atomic, unnested_dataframe)
 
-  reconstruct_tibble(data, out)
+  # Restore column order
```
It just occurred to me to wonder if it might be simpler to this a bit further up, i.e.:

```R
out <- c(rest, unnested_atomic, unnested_dataframe)
# fix order
out <- dplyr::bind_cols(out)
reconstruct_tibble(data, out)
```

aaronwolen replied:
We can capture the original order/structure of column names immediately, so the `nested_vars` object could be created much earlier in the function. But the actual reordering requires the output of `bind_cols(rest, unnested_atomic, unnested_dataframe)`, so I'm not sure that bit can be relocated.

aaronwolen replied:
Oh... 💡!

You meant reconstruct the tibble first and *then* take care of reordering the columns!

Something like this?

```r
  rest <- data[rep(seq_nrow(data), n[[1]]), group_vars, drop = FALSE]
  out <- reconstruct_tibble(
    data,
    dplyr::bind_cols(rest, unnested_atomic, unnested_dataframe)
  )

  # Restore column order
  nested_vars <- imap(data, function(.x, .y) {
    if (is_empty(.x)) {
      return(.y)
    } else if (is.data.frame(.x[[1]])) {
      return(names(.x[[1]]))
    } else {
      return(.y)
    }
  })
  ordered_vars <- intersect(unlist(nested_vars, use.names = FALSE), names(out))
  out[c(ordered_vars, setdiff(names(out), ordered_vars))]
```

hadley replied:
Oh yeah, maybe you're right and I'm confused about this structure.

aaronwolen commented on R/unnest.R:
```diff
@@ -157,7 +157,20 @@ unnest.data.frame <- function(data, ..., .drop = NA, .id = NULL,
+    } else {
+      return(.y)
+    }
+  })
```
I feel like this approach is easier to follow. Thoughts?

```r
  nested_vars <- imap(data, function(.x, .y) {
    if (is_empty(.x)) {
      return(.y)
    }
    .x <- .x[[1]]
    if (is.data.frame(.x)) {
      names(.x)
    } else {
      .y
    }
  })
```

aaronwolen replied:
Not at all! You gotta do what's best for the package and I'll just be happy to see the issue fixed. Either way, I'm always happy to help out where I can.
